proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -246,6 +259,17 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
 		}
 	}
 
+	// attempt is at least 1, therefore, failed attempts will be at least 0
+	reqInfo.FailedAttempts = attempt - 1
+
+	if reqInfo.FailedAttempts > 0 && lastFailedAttempt.IsZero() {
+		panic("forgot to record end of failed attempts")
+	}
+
+	if !lastFailedAttempt.IsZero() {
+		reqInfo.FailedAttemptsTime = lastFailedAttempt.Sub(beginAttempts).Seconds()
+	}
+
```

Comment:
why don't you replace this complexity with a simple
```
reqInfo.FailedAttempts++
reqInfo.FailedAttemptsTime = time.Now().Sub(beginAttempts).Seconds()
```
in the two cases?

Reply:
We considered this but it feels very brittle, we must increment the counter at different places and any future change must consider this. This way we rely on the attempts counter which is known to be correct and only modified in a single place. I'm also very unhappy with the way we set `lastFailedAttempt` because it suffers from the same issue (hence the panic to hopefully prevent such conditions by failing loud and clear during testing), not sure where I would put the second LOC you suggested.

Reply:
After further discussion we agreed to move add the two lines to each error case and remove the checks after the loop.